lib/pbench/server/cache_manager.py

```diff
@@ -173,6 +185,67 @@ def make_cache_object(dir_path: Path, path: Path) -> CacheObject:
     )
 
 
+class Inventory:
+    """Encapsulate the tarfile TarFile object and file stream
+
+    This encapsulation allows cleaner downstream handling, so that we can close
+    both the extractfile file stream and the tarball object itself when we're
+    done. This eliminates interference with later operations.
+    """
+
+    def __init__(self, stream: IO[bytes], tarfile: Optional[tarfile.TarFile] = None):
+        """Construct an instance to track extracted inventory
+
+        This encapsulates many byte stream operations so that it can be used
+        as if it were a byte stream.
+
+        Args:
+            stream: the data stream of a specific tarball member
+            tarfile: the TarFile object
+        """
+        self.tarfile = tarfile
+        self.stream = stream
+
+    def close(self):
+        """Close both the byte stream and, if open, a tarfile object"""
+        self.stream.close()
+        if self.tarfile:
+            self.tarfile.close()
+
+    def getbuffer(self):
+        """Return the underlying byte buffer (used by send_file)"""
+        return self.getbuffer()
+
+    def read(self, size: int = -1, /) -> bytes:
+        """Encapsulate a read operation"""
+        return self.stream.read(size)
+
+    def readable(self) -> bool:
+        """Return the readable state of the stream"""
+        return self.stream.readable()
+
+    def seek(self, offset: int, whence: int = SEEK_SET, /) -> int:
+        """Allow setting the relative position in the stream"""
+        return self.stream.seek(offset, whence)
+
+    def __str__(self) -> str:
+        """Return a string representation"""
+        return f"<Stream {self.stream} from {self.tarfile}"
```

Comment:
There's always a balance between `str()` and `repr()`. Either is often embedded in a string and a phrase can be easy to miss. I tend to offset the phrase in angle brackets rather than quotes or parens, and I often fall back to something much like the syntax of Java introspected `str` implementations. A canonical `repr` is usually designed to resemble (if not actually *be*) a constructor executable by `eval`... but, again, for embedding in a string I find the undelimited `name(args)` format awkward to read.

As for the `<` ... it's wearing a hat? 🤠

Reply:
I was keying off the leading `<` as being typical of a `__repr__()` output, e.g.:
```
>>> f = Path("/Users/wscales/.bashrc").open()
>>> f.__repr__()
"<_io.TextIOWrapper name='/Users/wscales/.bashrc' mode='r' encoding='UTF-8'>"
```
In this case, `TextIOWrapper` doesn't provide a `__str__()` implementation, so it falls back to `__repr__()`:
```
>>> f.__str__()
"<_io.TextIOWrapper name='/Users/wscales/.bashrc' mode='r' encoding='UTF-8'>"
```

What sort of object provides `name(args)` format?

Reply:
I've seen them, they're referenced generically in the documentation, and there's a school of thought that `repr` should be `eval`-able. E.g.,

>Return a string containing a printable representation of an object. For many types, this function makes an attempt to return a string that would yield an object with the same value when passed to [eval()](https://docs.python.org/3/library/functions.html#eval); otherwise, the representation is a string enclosed in angle brackets that contains the name of the type of the object together with additional information often including the name and address of the object

Maybe I'll just change this to `__repr__`.